src/storage/concurrency_manager/mod.rs

```diff
@@ -0,0 +1,144 @@
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
+
+mod lock_table;
+mod memory_lock;
+
+pub use self::memory_lock::{MemoryLock, TxnMutexGuard};
+
+use kvproto::kvrpcpb::LockInfo;
+use parking_lot::Mutex;
+use std::{
+    collections::BTreeMap,
+    mem::{self, MaybeUninit},
+    sync::{
+        atomic::{AtomicU64, Ordering},
+        Arc,
+    },
+};
+use txn_types::TimeStamp;
+
+type OrderedLockMap = Mutex<BTreeMap<Vec<u8>, Arc<MemoryLock>>>;
+pub type LockTable = self::lock_table::LockTable<OrderedLockMap>;
+
+#[derive(Clone)]
+pub struct ConcurrencyManager {
+    max_read_ts: Arc<AtomicU64>,
```

Comment:
Any advice on how we organize them?
My thought is to have something that can check locks for reading requests and update the max_read_ts at the same time. So I create this concurrent manager containing the lock table and the max_read_ts.

Reply:
I haven't thought deeply about it, but my intuition is that the lock/latch management part of things should be pretty well-encapsulated and deal only with managing concurrency with locks, etc. It should only touch lock CF, and even that should be easily abstracted. That way we can keep it fairly separate from the rest of the code, always run one per server, and more easily integrate it with scheduling or YATP or other concurrency stuff. Keeping track of timestamps etc, should be part of the higher level components (it could be dumped into `Storage`) so that we can easily change whether it is per-node/per-region/per-key or change the async commit algorithm, without touching the concurrency manager.

I don't have much to support this intuition, but one justification is that concurrency manager is going to grow lots of tricky, complex code (its already pretty complex) and it is best if we can test it in isolation from the rest of TiKV (not just unit tests, but ideally fuzz testing or simulation).

Reply:
`ConcurrencyManager` looks like a toolbox to me. So in my concept, `ConcurrencyManager` itself won't be complex, the complexity will be well encapsulated inside.

Any module which has effect to transactions owns a `ConcurrencyManager` and it can do what it wants.  For example, the raft store also needs to check locks when doing replica read, so it will also own one.

Reply:
> the raft store also needs to check locks when doing replica read, so it will also own one.

So, for raft store, it would reuse lock table and memory lock, but have a different concurrency manager?